rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+operation happens: creation, update, deletion of the resource. However, policies evolve over time.
+More policies can be added or they can be redeployed with different settings.
+As a result of that, a Kubernetes resource that was marked as valid even a couple of hours ago,
+could be instead rejected later on because of a policy change.
+
+We want to implement a audit check that inspects all the resources that are already
+present inside of the cluster and flags all the ones that are violating at least one of
+the currently enforced policies.
+
+The results of these evaluations must be accessible by operators and end users
+in an easy way.
+
+## Examples / User Stories
+
+As a operator, I want to know if some resource in my cluster violates the latest policies.
+
+As a user, I want to know if some of resource that I own violates the cluster compliance policies.
+
+As a Kubernetes developer, I want access to the audit checks results to allow me pro-grammatically act upon it.
+
+# Detailed design
+
+Kubewarden will generate audit reports using the
+[PolicyReport](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report)
+format that is standardized by the Kubernetes SIG "wg-policy".
+
+The audit reports are defined as set of Kubernetes CRDs. The CRDs are pretty
+flexible, they can be used to store audit
+results from Kubernetes policy engines (like Kubewarden or Kyverno) but can
+also be used to record audit results from other kind of tools (like Kubernetes
+security scanning tools as Falco or Kubebench).
+
+Kubewarden will generate one `PolicyReport` object per each Namespace found inside of
+the cluster.
+Plus, it will maintain one single `ClusterPolicyReport` object to keep track
+of all the cluster-wide Kubernetes resources.
+
+## Namespace report
+
+A Namespace can be "affected" both by `ClusterAdmissionPolicy` and by `AdmissionPolicy` policies.
+When auditing a Namespace, only some Kubewarden policies are going to be interested
+about the events happening inside of that Namespace. In addition to that,
+the same Kubernetes resource could be evaluated by multiple policies. This will
+lead to several policy evaluations being performed.
+
+The `PolicyReport` is a namespaced CR that provides these information:
+
+* How many policy evaluations have been done during the audit run of the Namespace
+* How many policy evaluations led to a resource being accepted, rejected or
+  could not be performed because of an error.
+  Policy failures happen when an error occurs during the policy evaluation,
+  for example, because of a runtime panic
+* Detailed results about all the evaluations performed. An evaluation result contains
+  the following information: the policy being used, the resource that was
+  evaluated, the final outcome (accept/reject/failure) and the eventual message
+  returned by the policy
+
+The `PolicyReport` of the Namespace is kept updated during the lifetime of the
```

Comment:
What if we don't keep updating the `PolicyReport` of the namespace? When a new audit check is triggered we could replace the old `PolicyReport` with the new one. 
If we keep adding new evaluations to the same `PolicyReport` I'm a bit concern we might end up with a big `PolicyReport` stored in etcd (it has limitations as you already mentioned).  It would also simplify the code.

We can assume a `PolicyReport` for a namespace is something that happened at some point with the state of the cluster at this time. And keep always the report for the latest check

We could keep some old reports instead of deleting them

Reply:
I like this idea. It makes also trivial to know if the ns/cluster is sane; if there's an empty "audit" `PolicyReport`, all is fine. If checks are to be run, one sees the old PolicyReport with old `PolicyReportResult` timestamps.

I would expect to have another set of `PolicyReports` and `ClusterPolicyReports`, that would contain only results from "monitor/enforcing" policies, and new requests. These "monitor/enforcing" reports would provide instant feedback if a policy is rejecting or approving requests, and these could contain an array of `PolicyReportResults` that is queue of reports, and that is garbage collected slow enough that the state of those resources would be present on the latest "audit" `PolicyReport`.

Reply:
It's up to us to decide how we are going to leverage the `PolicyReports` CR. However, other tools made for kyverno will take certain assumptions, if we do not respect these assumptions the results might be odd.

The main assumption is that a `PolicyReport` contains information about about all the evaluations done against the resources that were inside of a Namespace at a certain time. It is not tracking only the failures. By tracking also successful evaluations two main goals are achieved:

  * As a user I know with absolute certainty the compliance status of all the resources inside of a Namespace. If the PolicyReport doesn't contain any information about k8s resource X, that means the background checks has not evaluated it yet.
  * As Kubewarden authors, our code will know policy `polX` is interested about k8s reource `podA`. The code can look into the `PolicyReport` and look for an entry about `polX` evaluating `podA`. If no entry is found, the evaluation takes place. If the entry is found, we can look at the evaluation timestamp and compare it against the modification timestamps of `polX` and `podA`; if nothing changed since the last evaluation result we can just skip it and save time.

So the `PolicyReport` will contain only relevant information inside of it:
  * evaluations that are still valid, Old evaluation results are kept only when the policy AND the resource have not been changed since the evaluation time. This is done to reduce the number of evaluations we have to perform.
  * evaluations about resources/policies that have been deleted are removed

Despite that, I think there's the risk of having a big fat `PolicyReport` when a Namespace is crowded by k8s resources and/or many policies are enforced.

Reply:
Those are good points. I like the idea of not redoing evalutation for resources that haven't changed. 
I was just trying to think of ways to reduce the size of the `PolicyReport`. However it will probably just be ok since other tools are already using this approach and it looks like this is not a big issue.

It would be better to stick with the same approach, so we can use tools that are already made for `PolicyReport` (e.g. policy reporter)

Reply:
Do we have some stats from CNCF to let us know the average size of the namespaces? I'm wondering how big is the namespaces. Maybe this etcd size constraint is not a big problem now and we can iterate over it in the future. 

I see we wondering if the size of the report will be an issue. For that, I think we need to add more data in the discussion. Like, what's the limit evaluations number in the reports to get a good performance? In other words, how many result can we add to the report before seeing a performance degradation? Is there an easy way to check this?